pkg/tfbridge/info_defaults.go

```diff
@@ -0,0 +1,180 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tfbridge
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"unicode"
+
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	multierror "github.com/hashicorp/go-multierror"
+)
+
+// SetDefaults populates mappings such as the Resources mapping by computing sensible defaults in code. This is
+// especially useful for larger providers that have thousands of entries in Resources. Manually specified mappings
+// always take precedence: SetDefaults will never override an entry in the Resources map, only add new entries.
+func (info *ProviderInfo) SetDefaults(defaults ProviderInfoDefaults) error {
+	var errs multierror.Error
+	if defaults.Resources != nil {
+		if info.Resources == nil {
+			info.Resources = map[string]*ResourceInfo{}
+		}
+
+		if err := setDefaults(info.P.ResourcesMap(), info.Resources, defaults.Resources); err != nil {
+			errs.Errors = append(errs.Errors, fmt.Errorf("resources:\n%w", err))
+		}
+	}
+	if defaults.DataSources != nil {
+		if info.DataSources == nil {
+			info.DataSources = map[string]*DataSourceInfo{}
+		}
+		if err := setDefaults(info.P.DataSourcesMap(), info.DataSources, defaults.DataSources); err != nil {
+			errs.Errors = append(errs.Errors, fmt.Errorf("datasources:\n%w", err))
+		}
+	}
+	return errs.ErrorOrNil()
+}
+
+type ProviderInfoDefaults struct {
+	Resources   func(resourceToken string) (*ResourceInfo, error)
+	DataSources func(dataSourceToken string) (*DataSourceInfo, error)
+}
+
+func (d ProviderInfoDefaults) Unmappable(substring string) ProviderInfoDefaults {
+	return ProviderInfoDefaults{
+		Resources:   ensureUnmappable(substring, d.Resources),
+		DataSources: ensureUnmappable(substring, d.DataSources),
+	}
+}
+
+func ensureUnmappable[T any](
+	substring string,
+	f func(t string) (*T, error),
+) func(t string) (*T, error) {
+	return func(t string) (*T, error) {
+		if strings.Contains(t, substring) {
+			return nil, fmt.Errorf("token %q matches a substring %q that was specified as Unmappable",
+				t, substring)
+		}
+		return f(t)
+	}
+}
+
+type TFToken struct {
+	Module string
+	Name   string
+}
+
+type TFTokenParser func(tfToken string) (*TFToken, error)
+
+func SingleModule(prefix, defaultModule string) TFTokenParser {
+	return KnownModules(prefix, defaultModule, nil)
+}
+
+func KnownModules(prefix, defaultModule string, modules []string) TFTokenParser {
```

Comment:
So it looks to me that this is the essence of SingleModule/KnownModule strategies in the base branch. It tries to guess "terraform module" from the token, then feeds the result into standard token manglers. Perhaps this is useful then as its own mini-abstraction?

I'm thinking that on the base branch it's pre-packaged into types that carry ResourceInfo, when it's really only ever dealing with tokens, it feels a bit more natural to unpack it to only work on tokens, but YMMV it's not a big deal; the user can always compensate with a few more lambdas.

Reply:
I wanted it to work on `ResourceInfo` because I'm planning to follow up with support for stateful remappings (which benefit from alias support).

Reply:
Ah interesting. So you will need strategy combinators for that probably. I'd like to see that before taking a look at the naming again!